### PR TITLE
do not call finalize after exception

### PR DESCRIPTION
**Pull request by CheSema**
StorageS3Sink handle exceptions in a wrong way.
test_upload_s3_fail_create_multi_part_upload -- showed that object is tried to be written at onException method after exception
test_upload_s3_fail_upload_part_when_multi_part_upload -- the same, but it ends up as a FATAL error

test_heavy_insert_select_check_memory -- just good test, let it be

Continuation for https://github.com/ClickHouse/ClickHouse/issues/50274
Blocked by https://github.com/ClickHouse/ClickHouse/pull/50395

### Changelog category (leave one):
- Not for changelog (changelog entry is not required)

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

### CODE REVIEW

**Comment by robot-clickhouse-ci-2** [comment]
This is an automated comment for commit b37c00924dba856280c119d9ae32f80fb37dc89f with description of existing statuses. It's updated for the latest CI running
The full report is available [here](https://s3.amazonaws.com/clickhouse-test-reports/50907/b37c00924dba856280c119d9ae32f80fb37dc89f/ci_running.html)
The overall status of the commit is 🟡 pending

<table><thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody><tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td>🟡 pending</td></tr>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td>🟢 success</td></tr>
</table>

File: src/Storages/StorageS3.cpp [hunk 1]
```diff
@@ -794,7 +794,7 @@ class StorageS3Sink : public SinkToStorage
     void onException() override
     {
         std::lock_guard lock(cancel_mutex);
-        finalize();
+        abort();
```
Comment (alexey-milovidov) [inline src/Storages/StorageS3.cpp:797]: Ok. I was scared reading this function name, because it is the same as C's abort. Maybe `cancel`?

**Review by alexey-milovidov — APPROVED** [review]

File: src/Storages/StorageS3.cpp (hunk 1 above)
Comment (CheSema) [reply src/Storages/StorageS3.cpp:797]: `void onCancel()` is already in use.
Actually it is better to fix `onCancel` as well. But I do not have a test for it. And it is not urgent at all if compare to the changes in `onException`.

Comment (CheSema) [reply src/Storages/StorageS3.cpp:797]: May be `release` is fine here.

File: src/Storages/StorageS3.cpp [hunk 2]
```diff
@@ -824,6 +824,15 @@ class StorageS3Sink : public SinkToStorage
         }
     }
 
+    void release()
+    {
+        if (!writer)
+            return;
+
+        writer.reset();
+        write_buf.reset();
```
[the remark this answers is not in the document]
Comment (CheSema) [reply src/Storages/StorageS3.cpp:833]: Yes. I will add such test. Thank you.

File: src/IO/Lz4DeflatingWriteBuffer.cpp [hunk 3]
```diff
@@ -104,8 +104,14 @@ void Lz4DeflatingWriteBuffer::nextImpl()
             if (LZ4F_isError(compressed_size))
                 throw Exception(
                     ErrorCodes::LZ4_ENCODER_FAILED,
-                    "LZ4 failed to encode stream. LZ4F version: {}",
-                    LZ4F_VERSION);
+                    "LZ4 failed to encode stream. LZ4F version: {}, CodeName: {},"
```
Comment (CheSema) [inline src/IO/Lz4DeflatingWriteBuffer.cpp:107]: will revert this lines.

File: src/Storages/StorageS3.cpp [hunk 4]
```diff
@@ -813,10 +813,18 @@ class StorageS3Sink : public SinkToStorage
...
-        finalize();
+        try
+        {
+            std::rethrow_exception(exception);
+        }
+        catch(...)
+        {
+            /// An exception context is needed to proper delete write buffers without finalization
+            release();
+        }
```
Comment (Avogar) [inline src/Storages/StorageS3.cpp:827]: I think we should do the same for other storages that work with files like `StorageFile/StorageURL/StorageHDFS/StorageAzureBlob`.

**Review by Avogar — APPROVED** [review]

**Comment by CheSema** [comment]
Stateless tests (release)
01193_metadata_loading -- https://github.com/ClickHouse/ClickHouse/issues/51185

**Pull request merged**